subsys/usb/class/bt_h4.c

```diff
@@ -0,0 +1,315 @@
+#endif
+
+#if defined(CONFIG_BT_CTLR_TX_BUFFER_SIZE)
+#define BT_L2CAP_MTU (CONFIG_BT_CTLR_TX_BUFFER_SIZE - BT_L2CAP_HDR_SIZE)
+#else
+#define BT_L2CAP_MTU 64
+#endif
+
+/* Data size needed for CMD and ACL buffers */
+#define BUF_SIZE MAX(BT_BUF_RX_SIZE, BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU))
+NET_BUF_POOL_DEFINE(rx_pool, BUF_COUNT, BUF_SIZE, sizeof(u8_t), NULL);
+
+#define BT_H4_OUT_EP_ADDR               0x01
+#define BT_H4_IN_EP_ADDR                0x81
+
+#define BT_H4_OUT_EP_IDX                0
+#define BT_H4_IN_EP_IDX                 1
+
+#define BT_H4_BULK_EP_MPS               MIN(BUF_SIZE, USB_MAX_FS_BULK_MPS)
+
+/* HCI RX/TX threads */
+static K_THREAD_STACK_DEFINE(rx_thread_stack, 512);
+static struct k_thread rx_thread_data;
+static K_THREAD_STACK_DEFINE(tx_thread_stack, 512);
+static struct k_thread tx_thread_data;
+
+struct usb_bt_h4_config {
+	struct usb_if_descriptor if0;
+	struct usb_ep_descriptor if0_out_ep;
+	struct usb_ep_descriptor if0_in_ep;
+} __packed;
+
+USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_bt_h4_config bt_h4_cfg = {
+	/* Interface descriptor 0 */
+	.if0 = {
+		.bLength = sizeof(struct usb_if_descriptor),
+		.bDescriptorType = USB_INTERFACE_DESC,
+		.bInterfaceNumber = 0,
+		.bAlternateSetting = 0,
+		.bNumEndpoints = 2,
+		.bInterfaceClass = CUSTOM_CLASS, /* TBD */
+		.bInterfaceSubClass = 0,
+		.bInterfaceProtocol = 0,
+		.iInterface = 0,
+	},
+
+	/* Data Endpoint OUT */
+	.if0_out_ep = {
+		.bLength = sizeof(struct usb_ep_descriptor),
+		.bDescriptorType = USB_ENDPOINT_DESC,
+		.bEndpointAddress = BT_H4_OUT_EP_ADDR,
+		.bmAttributes = USB_DC_EP_BULK,
+		.wMaxPacketSize = sys_cpu_to_le16(BT_H4_BULK_EP_MPS),
+		.bInterval = 0x01,
+	},
+
+	/* Data Endpoint IN */
+	.if0_in_ep = {
+		.bLength = sizeof(struct usb_ep_descriptor),
+		.bDescriptorType = USB_ENDPOINT_DESC,
+		.bEndpointAddress = BT_H4_IN_EP_ADDR,
+		.bmAttributes = USB_DC_EP_BULK,
+		.wMaxPacketSize = sys_cpu_to_le16(BT_H4_BULK_EP_MPS),
+		.bInterval = 0x01,
+	},
+};
+
+#define H4_CMD 0x01
+#define H4_ACL 0x02
+#define H4_SCO 0x03
+#define H4_EVT 0x04
+
+static struct usb_ep_cfg_data bt_h4_ep_data[] = {
+	{
+		.ep_cb = usb_transfer_ep_callback,
+		.ep_addr = BT_H4_OUT_EP_ADDR,
+	},
+	{
+		.ep_cb = usb_transfer_ep_callback,
+		.ep_addr = BT_H4_IN_EP_ADDR,
+	},
+};
+
+static void bt_h4_read(u8_t ep, int size, void *priv)
+{
+	struct net_buf *buf = priv;
+
+	if (size > 0) {
+		buf->len += size;
+		net_buf_put(&rx_queue, buf);
+		buf = NULL;
+	}
+
+	if (buf) {
+		net_buf_unref(buf);
+	}
+
+	buf = net_buf_alloc(&rx_pool, K_FOREVER);
+
+	net_buf_reserve(buf, BT_BUF_RESERVE);
+
+	/* Start a new read transfer */
+	usb_transfer(bt_h4_ep_data[BT_H4_OUT_EP_IDX].ep_addr, buf->data,
+		     BUF_SIZE, USB_TRANS_READ, bt_h4_read, buf);
+}
+
+static void usb_h4_send(struct net_buf *buf)
+{
+	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_IN:
+		net_buf_push_u8(buf, H4_ACL);
```

Comment:
This is the BT_BUF_RESERVE, Right? This is now coming from the sample prj.conf. Could we change this dependency?

Reply:
Yes, I might change it to use select.